### DISCUSSION TITLE
TShock 1.4.x Full Release when?

### DISCUSSION THREAD

**Discussion by HiddenStriker** [Pryaxis/TShock#1839] [General]
First off, the title is very entitled, but I only mean it because I'm so excited for the full release of t-shock. I'm one of those that are waiting for a stable release to continue hosting a tshock server into 1.4. I love how fast you guys are putting out pre-releases and fixes, and am following your progress on twitter, and I know you said "a couple of weeks max for a stable release" before terraria 1.4 was out, but now since terraria 1.4.x is out and you're further into development, maybe we could get a more narrowed down release date? I hope it's soon!

Great work guys, and I'm looking forward to the stable and full version of tshock!

**Comment by HiddenStriker** [comment]
I guess this means its gunna be a long time :(

**Comment by hakusaro** [comment]
First of all, you can track our progress on total support for a stable 4.4.0 release here: https://github.com/Pryaxis/TShock/projects/4.

## What constitutes a stable release?

A stable release of TShock is something we consider "good enough" to run a public server on with reasonable expectations about. This means that the world's most naive hacked client shouldn't be able to case catastrophic world damage quite easily. This also means that all of the new content works properly under anticheat conditions, and that any of the TShock-exclusive features (like SSC) are working fully as intended and are mostly free of screwups. Right now, we need work on all three areas.

### Vanilla content

Terraria 1.4.0.4 has a lot of content that's destructive -- namely explosives. All of the new content needs to be supported. Currently, we're getting close to supporting all vanilla content. We're only a few items off.

This doesn't include perfecting vanilla content, for example. Spawn timers are something we still need to work on, as well as beds. Neither of these are trivial for a lot of reasons, but we're working on them. We don't think that it's very useful for us to focus on fixing spawn timers right now, for example, because it's mostly not a difference if you're dead for 7 or 30 seconds. Same with beds -- honestly you can solve almost all of the bed functionality with warps and we've seen most of the bed stuff working okay.

### Exclusive features

Kick on hardcore death and SSC are part of the core feature set that we need to have working. Neither are fully functional right now. SSC is a somewhat special feature -- Re-Logic implemented it in the client for us, but we need to take additional precautions to make sure it works. Right now, there are some bugs with it we need to sort out (like how journey mode interacts with it).

### Anti-crash

TShock protects against crashes that hacked clients can try. Currently, we think we're okay here, but not fully sure.

### Anti-cheat

This is protection against cheating for the sake of scumming gameplay. Right now, it's extremely easy to cheat on TShock servers. TShock is built first and foremost for large scale multiplayer, not small scale multiplayer. So we fix for small scale first, but larger servers expect us to have basic anti-cheat enabled.

### Anti-grief

TShock also contains basic anti-grief that helps stop your server from being trashed. This is pretty important if you port forward your server, so we're working on it too.

### APIs

We provide a number of events for other plugins. We haven't yet supported all of the new events, and we need to.
 
## In summary

In order for us to release TShock 4.4.40 (stable) for Terraria 1.4.0.x, we need:

1. Vanilla content to be working.
2. Exclusive features to be working.
3. Anti-crash to be working.
4. Anti-cheat to be working.
5. Anti-grief to be working.
6. APIs to be finalized.

We're currently nearing completion of 1-3, with 4-5 in-progress (earlier than the others, and 6 totally not started. TShock is made by volunteers, and ideally you can thumbs up issues on GitHub if you want them to be worked on first. We know people really want smart doors working, so that's pretty high on our list, for example.

**Reply by HiddenStriker** [reply]
Whoa, I didn't know a -lot- of this, this clears things up quite well, thanks for taking the time for this extremely detailed response, and good luck to you guys again! You should know many people love your work and are super excited for it!
